### ISSUE TITLE
Enhanced styling based on request response life cycle events

### ISSUE THREAD

**Issue by sirinath** [bigskysoftware/htmx#116]
Can we have the ability to explicitly specify styling information based on request-response life cycle events using `hx-swap` and `hx-indicator`. E.g. in the case of `hx-indicator`

```html
<div>
    <button hx-post="/example" hx-indicator="#spinner
         style:htmx-indicator-common
         error-style:htmx-error
         before-request:htmx-indicator-pre-request-common
         requesting:htmx-request-common
         after-request:htmx-indicator-post-request-common
         before-request-init:htmx-indicator-pre-request-init
         requesting-init:htmx-requesting-init
         after-request-init:htmx-indicator-post-request-init
         before-request-odd:htmx-indicator-before-request-odd
         requesting-odd:htmx-requesting-odd
         after-request-odd:htmx-indicator-after-request-odd
         before-request-even:htmx-indicator-before-request-even
         requesting-even:htmx-requesting-even
         after-request-even:htmx-indicator-after-request-even">
        Post It!
    </button>
    <img  id="spinner" src="/img/bars.svg"/>
</div>
```

**Comment by sirinath** [comment]
Maybe this would be better:

```html
<div>
    <button hx-post="/example" hx-indicator="#spinner first:htmx-indicator-first requesting-first:htmx-request-first odd:htmx-indicator-odd requesting-odd:htmx-request-odd even:htmx-indicator-even requesting-even:htmx-request-even">
        Post It!
    </button>
    <img  id="spinner" src="/img/bars.svg"/>
</div>
```

Similarly for swap.

**Comment by bencroker** [comment]
Wouldn’t default behavior be better suited to configuration settings?
https://htmx.org/docs#config

**Comment by sirinath** [comment]
This is the only way then you can only have one option for all elements. If they can be specified independently then you can have different styles for different elements.

**Comment by sirinath** [comment]
Also, I am not sure if `hx-indicator` is the best name for this.

**Comment by benpate** [comment]
I'm with bencroker -- configuring this kind of thing is great, but if I had to enter all of that for every... single... button... it would be maddening.

I wouldn't consider it magic, just convention over configuration.  Besides, CSS gives developers plenty of ways to switch the visual appearance of a node without touching the node:

```
<div class="my-red-thing><span class="htmx-indicator"></span></div>
<div class="my-blue-thing"><span class="htmx-indicator"></span></div>

<style>
  .my-red-thing .htmx-indicator {color:red}
  .my-blue-thing .htmx-indicator {color:blue}
</style>
```

Let's not ask BigSky for another huge library again.  There are plenty of those already :)

**Comment by sirinath** [comment]
Well, what I am saying is the main feature of the library is the data request part through the HTTP request. So there should be an easy way to change styles bases on phases gone though when HTTP requests are made. Also, styles applied when retrieving data say for and image may not be relevant for text, table or form so one size does not fit all in this case. This is why I made the suggestion. 

To summarise:
- styles changes based on the data updating process through HTTP requests and other means
- different styles for different elements based on their use cases

**Comment by bencroker** [comment]
I don't think htmx should deal with styles, that is what CSS classes are for. Is there something specific you are trying to achieve that you cannot with `hx-indicator`?

**Comment by sirinath** [comment]
Essentially more control over what `hx-indicator` does and what CSS classes to apply base on what htmx is doing. Plain CSS does not know when htmx is going to retrieve data. Maybe you can do this with complex javascript hacks but this defeats the purpose of using htmx. When it comes to `hx-indicator`, at least you might want to have a particular style before data is updated for the 1st time, the 1st time it is updated, when the data is going to change, etc.

**Comment by bencroker** [comment]
Can you provide a real-world use-case so we can see the limitations you are describing?

**Comment by sirinath** [comment]
Something like:
- awaiting data - style 1 to show that it is awaiting data
- initial data - style 2 to show this us the initial data set
- modified data - style 3 to show this data has been updated
- if there is an error getting data, style 4 to show this element is error

**Comment by bencroker** [comment]
My approach would be to handle this in the returned HTML. 

- awaiting data: use `hx-indicator` class
- initial data: apply style 2 in the returned HTML
- modified data - apply style 3 in the returned HTML
- error: apply the error styles in the returned HTML

For the type of behaviour you are requesting, it seems that htmx provides you with all of the [events](https://htmx.org/reference#events) you need and that it can be accomplished with [hyperscript](https://hyperscript.org/), albeit with a bit of extra code.

**Comment by sirinath** [comment]
Issues here is if there is an error there might be no returned HTML. Also, the style is applied to an element other than the one being updated hence you cannot apply the style to retuned HTML.

**Comment by benpate** [comment]
Unless the error is a network error, then the server should be able to return something.  If it is a network error (case 4 in both examples), then I think [htmx:onLoadError](https://htmx.org/events/#htmx:onLoadError) would be a good place to start.  

HTMX might have a role to play in a "case 4" situation, but I'm not sure yet how exactly that should be handled declaratively.

**Comment by bencroker** [comment]
Paging @1cg for his thoughts on this.

**Comment by 1cg** [comment]
agreed, too elaborate for core htmx, but (I sound like a broken record) something an extension could implement to explore
